providers/writing_a_new_provider.md

```diff
@@ -0,0 +1,555 @@
+
+Now lets look at collecting inventory.  For that let's look at :shocked: the collector.
+
+The collector provides an interface for the parser, so each method should fetch and return the relevant inventory.
+
+```ruby
+class ManageIQ::Providers::AwesomeCloud::Inventory::Collector < ManageIQ::Providers::Inventory::Collector
+  require_nested :CloudManager
+
+  def images
+    compute_client.get_images
+  end
+
+  def instances
+    compute_client.get_instances
+  end
+
+  def instance_types
+    compute_client.get_instance_types
+  end
+
+  private
+
+  def compute_client
+    @compute_client ||= manager.connect(:service => "Compute")
+  end
+end
+```
+
+And that's it!  The collector gets a lot more interesting when you add support for targeted refresh but that is for another time.  If you have to manually handle paging you should do that here, if the sdk handles paging automatically via an Enumerator then there's nothing more needed.
+
+Now we can get started on the parser.
+
+```ruby
+class ManageIQ::Providers::AwesomeCloud::Inventory::Parser < ManageIQ::Providers::Inventory::Parser
+  require_nested :CloudManager
+
+  def parse
+    instance_types
+    images
+    instances
+  end
+
+  def instance_types
+    # Calling collector.flavors here is what actually issues the API call.
+    collector.instance_types.each do |instance_type|
+      # At this point "instance_type" will be whatever is returned by your SDK.
+      # It could be a hash or it could be an object like `AwesomeCloud::Compute::InstanceType`
+
+      # persister.flavors.build will create an InventoryObject (fancy hash) with all of the
+      # attributes that you pass in here, and automatically add it to the `persister.flavors`
+      # inventory collection
+      persister.flavors.build(
+        # MIQ typically uses "ems_ref" as the unique reference for an inventory item
+        # in a provider.  Whatever you use must be guaranteed to not have a duplicate
+        # in the same provider instance as this value is also used to lookup related
+        # inventory from other collections.
+        :ems_ref => instance_type.id,
+        :name    => instance_type.name,
+        :cpus    => instance_type.n_cpus,
+        :memory  => instance_type.ram
+      )
+    end
+  end
+
+  def images
+    collector.images.each do |image|
+      persister.miq_templates.build(
+        :ems_ref         => image.id,
+        # The uid_ems field if you see it typically indicates a field that can be used
+        # to identify an inventory item across provider instances.  Most of the time
+        # for cloud providers this is the same as the ems_ref but not always.
+        # If your ems_ref looks like an integer ID then it probably isn't unique
+        # If it looks like a UUID then it probably is.
+        :uid_ems         => image.id,
+        :name            => image.name,
+        :location        => "unknown",
+        :raw_power_state => "never",
+        :template        => true,
```

Comment:
`raw_power_state` and `template` are the same for every provider, consider adding these as default values to the core miq_templates inventory collection definition